allen-cell-animated/timelapse-colorizer
# feature: Improved load errors

Problem
=======
Closes #411, "Show more informative errors when datasets fail to load".

This change adds additional catches for errors and adds more informative messages that can be shown to users. It also adds explicit onscreen warnings for issues that were previously only logged to the console.

*Estimated size: medium, 30 minutes*

Solution
========
- Added callbacks for errors and warnings in the Viewer which can be passed to the Collection + Dataset classes during load operations.
- Added server error handlers in file loaders for 404 errors, etc.
- Datasets can now load even when the times data is missing. (This causes the scatterplot and plot views to not work correctly.)

New warnings and errors:
- Added warnings for missing and duplicated feature and data files.
- Added explicit error messages for missing properties in collection and dataset files.
- The load dialog now displays error messages for both collection and dataset files when the file type is ambiguous.

## Type of change
* New feature (non-breaking change which adds functionality)

Change summary:
---------------
* Tidy, well formulated commit message
* Another great commit message
* Something else I/we did

Steps to Verify:
----------------
Preview link: https://allen-cell-animated.github.io/timelapse-colorizer/pr-preview/pr-453

| Name | Test Steps | Expected behavior/error message | 
| -- | -- | -- |
| 404 error | 1. Open the preview link and open the load dialog. <br/>2. Paste in this URL: https://dev-aics-dtp-001.int.allencell.org/dan-data/colorizer/data/test/bad-data/no-exist | "Error: Could not load the provided URL as either a collection or a dataset: server returned a 404 (Not Found) code." |
| Empty collection | 1. Open the preview link and open the load dialog. <br/>2. Paste in this URL: https://dev-aics-dtp-001.int.allencell.org/dan-data/colorizer/data/test/bad-data/empty-collection.json | "The collection JSON was loaded but no datasets were found. At least one dataset must be defined in the collection." |
| Duplicate collection keys | 1. Open this link: https://allen-cell-animated.github.io/timelapse-colorizer/pr-preview/pr-453/viewer?collection=https%3A%2F%2Fdev-aics-dtp-001.int.allencell.org%2Fdan-data%2Fcolorizer%2Fdata%2Ftest%2Fbad-data%2Fcollection-with-duplicates.json | You should see a warning banner with the text "Duplicate dataset names were found in the collection." |
| Bad manifest when loading datasets | 1. Open this link: https://allen-cell-animated.github.io/timelapse-colorizer/pr-preview/pr-453/viewer?collection=https%3A%2F%2Fdev-aics-dtp-001.int.allencell.org%2Fdan-data%2Fcolorizer%2Fdata%2Ftest%2Fbad-data%2Fcollection.json&dataset=Malformed+JSON+manifest | An error banner should appear instead of a popup warning about the dataset load issue |

More interactive examples here: https://allen-cell-animated.github.io/timelapse-colorizer/pr-preview/pr-453/viewer?collection=https%3A%2F%2Fdev-aics-dtp-001.int.allencell.org%2Fdan-data%2Fcolorizer%2Fdata%2Ftest%2Fbad-data%2Fcollection.json

Screenshots (optional):
-----------------------


![image](https://github.com/user-attachments/assets/13066171-12f4-47d1-b84e-1dfdc4dda9da)


**Video talk-through (🔊):**

https://github.com/user-attachments/assets/039c6daf-974d-446a-9325-4a89c955fcab
